src/client/datascience/kernel-launcher/localKernelFinder.ts

```diff
@@ -122,17 +159,71 @@ export class LocalKernelFinder implements ILocalKernelFinder {
         ]);
 
         const kernels = this.filterKernels(nonPythonKernelSpecs.concat(pythonRelatedKernelSpecs));
-        sendKernelListTelemetry(resource, kernels);
+        this.lastFetchedKernelsWithoutCache = kernels;
+        this.globalState.update('JUPYTER_GLOBAL_KERNELSPECS', kernels).then(noop, (ex) => {
+            console.error('Failed to update global kernel cache', ex);
+        });
         return kernels;
     }
 
-    // This should return a WRITABLE place that jupyter will look for a kernel as documented
-    // here: https://jupyter-client.readthedocs.io/en/stable/kernels.html#kernel-specs
-    public async getKernelSpecRootPath(): Promise<string | undefined> {
-        return this.jupyterPaths.getKernelSpecRootPath();
+    private async listValidKernelsFromGlobalCache(
+        cancelToken?: CancellationToken
+    ): Promise<LocalKernelConnectionMetadata[]> {
+        const values = this.lastFetchedKernelsWithoutCache.length
+            ? this.lastFetchedKernelsWithoutCache
+            : this.globalState.get<LocalKernelConnectionMetadata[]>('JUPYTER_GLOBAL_KERNELSPECS', []);
+        const validValues: LocalKernelConnectionMetadata[] = [];
+        const promise = Promise.all(
+            values.map(async (item) => {
+                let somethingIsInvalid = false;
+                const promises: Promise<void>[] = [];
+                if (item.interpreter?.path) {
+                    // Possible the interpreter no longer exists, in such cases, exclude this cached kernel from the list.
+                    promises.push(
+                        this.fs
+                            .localFileExists(item.interpreter.path)
+                            .then((exists) => {
+                                if (!exists) {
+                                    somethingIsInvalid = true;
+                                }
+                            })
+                            .catch(noop)
+                    );
+                }
+                if (item.kind === 'startUsingKernelSpec' && item.kernelSpec?.specFile) {
+                    // Possible the kernelspec file no longer exists, in such cases, exclude this cached kernel from the list.
+                    promises.push(
+                        this.fs
+                            .localFileExists(item.kernelSpec.specFile)
```

Comment:
I'm not sure this is true? Do we ever create 'startUsingKernelSpec' without a specfile?

Reply:
As in I don't see this condition from before.

Reply:
The code in the jupyterKernelService seems to rely on the fact that the specFile doesn't exist.

Reply:
Its possible the user deleted the kernelspec file.

> As in I don't see this condition from before.

Thats correct, thats because we always serach and return only kernelspecs that exist on disc.
However now that I'm caching it, its possibel the user subsequently uninstalls/deletes that kernelspec from disc. In which case, the item in the cache is no longer valid (as it doesn't exist).

Reply:
So yes we only search for ones on disk, but I thought we generate ones too, where the file may not have ever existed. So this would destroy those (as well as removing the ones you intend).

Reply:
> here the file may not have ever existed. So this would destroy those (as well as removing the ones you intend).

In that case, aren't we starting from `PythonInterpreter` & not from kernel spec.

Reply:
I.e. if we're starting from a kernel spec, then the spec file must exist & this code ensures we only deal with those kernels where `kind= startUsingKernelSpec`

Reply:
> aren't we starting from PythonInterpreter & not from kernel spec.

I can't say what the type would end up as. I mean you can test it and find out. Or maybe just look at all the spots where it's 'startUsingKernelSpec' and verify they always need a kernelspec. I thought we always create a 'default' one that may or may not have an actual kernelspec on disk.

Reply:
> where it's 'startUsingKernelSpec' and verify they always need a kernelspec.

Yup, i checked, and can confirm that it must exist.

> thought we always create a 'default' one that may or may not have an actual kernelspec on disk.

Yes, in that case the `kind=DefaultKernelConnectionMetadata`
Hence the code i have should work correctly.

Reply:
> thought we always create a 'default' one that may or may not have an actual kernelspec on disk.

Yes, thats for when we start Python interpreters (we have a bogus kernelspec, which will get created only when we register the kernel), Again, i'm not interested in those (not interested when we start using Python interpreter).